core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -926,30 +933,49 @@ private[spark] class BlockManager(
         } else if (level.useDisk && diskStore.contains(blockId)) {
-          val diskData = diskStore.getBytes(blockId)
-          val iterToReturn: Iterator[Any] = {
-            if (level.deserialized) {
-              val diskValues = serializerManager.dataDeserializeStream(
-                blockId,
-                diskData.toInputStream())(info.classTag)
-              maybeCacheDiskValuesInMemory(info, blockId, level, diskValues)
-            } else {
-              val stream = maybeCacheDiskBytesInMemory(info, blockId, level, diskData)
-                .map { _.toInputStream(dispose = false) }
-                .getOrElse { diskData.toInputStream() }
-              serializerManager.dataDeserializeStream(blockId, stream)(info.classTag)
+          try {
+            val diskData = diskStore.getBytes(blockId)
+            val iterToReturn: Iterator[Any] = {
+              if (level.deserialized) {
+                val diskValues = serializerManager.dataDeserializeStream(
+                  blockId,
+                  diskData.toInputStream())(info.classTag)
+                maybeCacheDiskValuesInMemory(info, blockId, level, diskValues)
+              } else {
+                val stream = maybeCacheDiskBytesInMemory(info, blockId, level, diskData)
+                  .map { _.toInputStream(dispose = false) }
+                  .getOrElse { diskData.toInputStream() }
+                serializerManager.dataDeserializeStream(blockId, stream)(info.classTag)
+              }
             }
+            val ci = CompletionIterator[Any, Iterator[Any]](iterToReturn, {
+              releaseLockAndDispose(blockId, diskData, taskContext)
+            })
+            Some(new BlockResult(ci, DataReadMethod.Disk, info.size))
+          } catch {
+            case ex: KryoException if ex.getCause.isInstanceOf[IOException] =>
+              // We need to have clear error message to catch environmental problems easily.
+              // Further details: https://issues.apache.org/jira/browse/SPARK-37710
+              processKryoException(ex, blockId)
+              throw ex
           }
-          val ci = CompletionIterator[Any, Iterator[Any]](iterToReturn, {
-            releaseLockAndDispose(blockId, diskData, taskContext)
-          })
-          Some(new BlockResult(ci, DataReadMethod.Disk, info.size))
         } else {
           handleLocalReadFailure(blockId)
         }
     }
   }
 
+  private def processKryoException(ex: KryoException, blockId: BlockId): Unit = {
+    var errorMessage = s"${ex.getMessage}. Please check if environment status is healthy " +
+      s"(e.g: disk corruption, network failure (etc)). " +
+      s"${blockManagerId.toString} - blockName: $blockId"
+    if (diskBlockManager.containsBlock(blockId)) {
+      val file = diskBlockManager.getFile(blockId)
+      errorMessage = errorMessage + s" - blockDiskPath: ${file.getAbsolutePath}"
+    }
+    logError(errorMessage)
```

Comment:
I probably mentioned this already, change this to `logDebug` or `logInfo`.

Reply:
`logDebug`: `INFO` log level is mostly used at production systems so for example: in production, if disk corruption occurs on Kryo deserialization level, `problematic BlockManager, blockId` and `blockPath` information will be missed with `debug` log level.
`logInfo`: This log message covers `ex.getMessage` and is going to be printed in only `error` case for `problematic BlockManager, blockId` and `blockPath` information.

In the light of these, i could not see a good reason to be used `info` or `debug` level for this use-case.

Reply:
At least  warn seems appropriate. If it's possibly recoverable, that's fine. If not, error.

Reply:
Exactly, we can use `error/warn` in this case, however, not `info/debug`.

This kind of problems are retried with `task-attempts` (3 attempts in our case). 
- If I/O issue is `transient`, it can be recovered by next task-attempt. 
- If it is not transient, it will fail `task/stage/job/query` (after all attempts).
I think we can also use `warn` due to current task-retry flow (even if the first task attempt fails, flow will re-try until retry-limit).

Reply:
The actual failure is logged/handled where the `KryoException` is caught - this log message is to give additional information to help debug in case someone wants to understand more on why the KryoException is thrown.
As such, this is informational at best and debug if it becomes very verbose - I am err'ing on side of caution that it makes sense to make this INFO.

For example, we have had very verbose error/warn messages in past during executor shutdown (deletion of directories/files which cause a bunch of IOExceptions), which end up with polluting the logs.

Reply:
I think the main goal of this patch is to provide `problematic BlockManager`, `blockId` and `blockPath` (in addition `error` message) when specially `disk problems` occurred on `prod` systems (using `INFO` log level).
For example, our use-case was `disk corruption` as a real-life example and if `problematic BlockManager`, `blockId` and `blockPath` are logged at DEBUG level where prod systems mostly run at INFO log-level, these hints will be missed on prod logs. On the other hand, it is hard to catch this kind of environmental problems on prod systems and also, hard to reproduce on test environments (unless additional test pipeline is built to simulate network conditions, e.g: `toxiproxy`).

I updated this log message level from ERROR to WARN and i think it is useful to be logged explicitly on prod systems if environmental problems (e.g: disk corruption) are occurred.

Reply:
I am not in favor of making this ERROR or WARN - it is not actionable for end users, ends up surfacing for a larger number of cases (not just unrecoverable errors), and the actual message is handled elsewhere while this PR giving supporting information to help triage an issue.

I understand that there some specific set of corner cases where this information would be useful (like disk corruption, etc) - which is why I am fine with leaving it at INFO instead of DEBUG.

Reply:
I disagree it's unactionable and seems important to surface; I don't think it's one of these things that can happen 'normally' and ends up being log noise. But OK let's put it at INFO to close this out

Reply:
I gave an example of why it is unactionable - when executors shutdown, we end up with files getting deleted, and IOException's getting thrown in this path.